src/liballoc/collections/btree/node.rs

```diff
@@ -959,11 +1031,12 @@ impl<BorrowType, K, V> Handle<NodeRef<BorrowType, K, V, marker::Internal>, marke
     /// `edge.descend().ascend().unwrap()` and `node.ascend().unwrap().descend()` should
     /// both, upon success, do nothing.
     pub fn descend(self) -> NodeRef<BorrowType, K, V, marker::LeafOrInternal> {
+        // We need to use raw pointers to nodes because, if BorrowType is marker::ValMut,
+        // there might be outstanding mutable references to values that we must not invalidate.
+        let internal_node = self.node.as_internal();
         NodeRef {
             height: self.node.height - 1,
-            node: unsafe {
-                (&*self.node.as_internal().edges.get_unchecked(self.idx).as_ptr()).as_ptr()
-            },
+            node: unsafe { (&*(*internal_node).edges.get_unchecked(self.idx).as_ptr()).as_ptr() },
```

Comment:
This `&*` of the pointer contained by MaybeUninit is something inspiring my comment in #73987 - as far as I understand, it causes the `BoxedNode` coming out to a stand-up citizen. I have no idea what for. I have previously read a post/article about the purpose of `&*` but can't find it back.

Reply:
The `&*` here applies to `(*internal_node).edges.get_unchecked(self.idx).as_ptr()`. So really this is just
```rust
(*internal_node).edges.get_unchecked(self.idx).as_ptr() as &_
```
except you cannot cast to a shard reference.

https://github.com/rust-lang/rust/issues/73987 is about the fact that your code really is short for
```rust
...::get_unchecked((*internal_node).edges as &_, self.idx).as_ptr() as &_
```
(again using my pseudo-casts). It is the first, inner "cast" I am worried about, not the second one.

Reply:
I didn't even think about the edge access brining up this line. You're saying that during `iter_mut`, a descend will still stake its claim on the entire edge array? But since we don't hand out references to edges, that's not actually a problem. Probably. Yet. It did inspire me for [another playground example of UB with a more subtle way of clinging on to references](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=8319fe76c5ae23469f3f50564b4ef0ce), but that no longer fails with this PR.

Reply:
I think I understand the `&*` (that existed in the old code) now: it's mainly `*`, illustrating that a auto-ref is done for the call to `BoxedNode::as_ptr(&self)`, and if you try to figure this out with intermediate variable and types the wrong way, you can end up with a basically superfluous. `&*`.

Reply:
> You're saying that during iter_mut, a descend will still stake its claim on the entire edge array?

Yes. And yes that seems fine as the structure of the tree does not change (hence why `ValMut` is so important), but it is worth a comment.

> It did inspire me for another playground example of UB with a more subtle way of clinging on to references, but that no longer fails with this PR.

You are getting good at coming up with these examples. ;) Another one for the test suite!

Reply:
Comment added (not pushed yet) and an example of easier UB than my first example added. I haven't investigated why simply moving a second (reverse) iterator invalidates key/value references in stable/nightly, but I guess it's reading the length of a node just to create a handle.